[Driver] Fix _XOPEN_SOURCE definition on Solaris

rorth opened the pull request:
Since commit 613a077b05b8352a48695be295037306f5fca151, `flang` doesn't build any longer on Solaris/amd64:
```
flang/lib/Evaluate/intrinsics-library.cpp:225:26:
error: address of overloaded function 'acos' does not match required type '__float128 (__float128)'
  225 |       FolderFactory<F, F{std::acos}>::Create("acos"),
      |                          ^~~~~~~~~
```
That patch led to the version of `quadmath.h` deep inside `/usr/gcc/<N>` to be found, thus `HAS_QUADMATHLIB` is defined.  However, the `struct HostRuntimeLibrary<__float128, LibraryVersion::Libm>` template is guarded by `_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600`, while `clang` only predefines `_XOPEN_SOURCE=500`.

This code dates back to commit 0c1941cb055fcf008e17faa6605969673211bea3 back in 2012.  Currently, this is long obsolete and `gcc` prefefines `_XOPEN_SOURCE=600` instead since GCC 4.6 back in 2011.

This patch follows that.

Tested on `amd64-pc-solaris2.11` and `sparcv9-sun-solaris2.11`.

rorth committed:
[Driver] Fix _XOPEN_SOURCE definition on Solaris

Since commit 613a077b05b8352a48695be295037306f5fca151, `flang` doesn't
build any longer on Solaris/amd64:
```
flang/lib/Evaluate/intrinsics-library.cpp:225:26:
error: address of overloaded function 'acos' does not match requi
red type '__float128 (__float128)'
  225 |       FolderFactory<F, F{std::acos}>::Create("acos"),
      |                          ^~~~~~~~~
```
That patch led to the version of `quadmath.h` deep inside `/usr/gcc/<N>` to
be found, thus `HAS_QUADMATHLIB` is defined.  However, the `struct
HostRuntimeLibrary<__float128, LibraryVersion::Libm>` template is guarded
by `_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600`, while `clang`
only predefines `_XOPEN_SOURCE=500`.

This code dates back to commit 0c1941cb055fcf008e17faa6605969673211bea3
back in 2012.  Currently, this is long obsolete and `gcc` prefefines
`_XOPEN_SOURCE=600` instead since GCC 4.6 back in 2011.

This patch follows that.

Tested on `amd64-pc-solaris2.11` and `sparcv9-sun-solaris2.11`.

MaskRay reviewed (approved):
(I have limited internet access between April 20th and May 4th, and my response time may be delayed..)

rorth commented:
/cherry-pick e71c8ea3cc73c8f7b0382468f355a254166d3a72

tstellar commented:
Was there a backport PR created for this change?

rorth commented:
No.  However, I got an [error](https://github.com/llvm/llvm-project/actions/runs/14798017932) for the `/cherry-pick` that looks like some weird internal error to me, not a failed merge attempt or some such.

rorth commented:
Drats, I'd hoped the markup would inhibit the cherry pick directive from being picked up.

I've now dropped the patch into local LLVM 20.1.5 builds on both `amd64-pc-solaris2.11` and `sparcv9-sun-solaris2.11`.  It applied cleanly and introduced no regressions.  I'll try a regular cherry pick again: maybe it finally works this time.

rorth commented:
/cherry-pick e71c8ea3cc73c8f7b0382468f355a254166d3a72
